File: nofishing-backend/src/main/java/com/nofishing/entity/ApiKey.java
package com.nofishing.entity;

import jakarta.persistence.*;
import lombok.Data;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

/**
 * API Key entity for external API access
 */
@Entity
@Table(name = "api_key")
@Data
@EntityListeners(AuditingEntityListener.class)
public class ApiKey {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true, nullable = false, length = 64)
    private String keyValue;

    @Column(nullable = false, length = 100)
    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(length = 500)
    private String permissions;  // JSON: ["DETECT", "HISTORY_READ", "HISTORY_DELETE"]

    private LocalDateTime expiresAt;

    private LocalDateTime lastUsedAt;

    @Column(nullable = false)
    private Boolean isEnabled = true;

    @Column(length = 50)
    private String createdBy;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    /**
     * Permission types
     */
    public enum Permission {
        DETECT,
        HISTORY_READ,
        HISTORY_DELETE,
        WHITELIST_READ,
        WHITELIST_WRITE,
        BLACKLIST_READ,
        BLACKLIST_WRITE,
        STATS_READ
    }

    /**
     * Check if key is expired
     */
    public boolean isExpired() {
        return expiresAt != null && LocalDateTime.now().isAfter(expiresAt);
    }

    /**
     * Check if key is valid (enabled and not expired)
     */
    public boolean isValid() {
        return isEnabled && !isExpired();
    }
}

File: nofishing-dashboard/src/api/apiKey.ts
import { apiClient } from './client';

export interface ApiKey {
  id: number;
  keyValue: string;
  name: string;
  userId: number;
  username?: string;
  permissions: string[];
  expiresAt?: string;
  lastUsedAt?: string;
  isEnabled: boolean;
  createdBy?: string;
  createdAt?: string;
}

export interface CreateApiKeyRequest {
  name: string;
  userId: number;
  permissions: string[];
  expiresAt?: string;
}

export const apiKeyApi = {
  // Create API key
  createKey: (data: CreateApiKeyRequest) =>
    apiClient.post<ApiKey>('/api/v1/admin/api-keys', data),

  // Get all keys
  getAllKeys: () =>
    apiClient.get<ApiKey[]>('/api/v1/admin/api-keys'),

  // Get keys by user
  getUserKeys: (userId: number) =>
    apiClient.get<ApiKey[]>(`/api/v1/admin/api-keys/user/${userId}`),

  // Get key by ID
  getKeyById: (id: number) =>
    apiClient.get<ApiKey>(`/api/v1/admin/api-keys/${id}`),

  // Revoke key
  revokeKey: (id: number) =>
    apiClient.delete<void>(`/api/v1/admin/api-keys/${id}`),

  // Disable key
  disableKey: (id: number) =>
    apiClient.post<void>(`/api/v1/admin/api-keys/${id}/disable`),

  // Cleanup expired keys
  cleanupExpiredKeys: () =>
    apiClient.post<void>('/api/v1/admin/api-keys/cleanup'),
};
